# Handwritten digits Classifier

Exercise

Use-> Logistic Regression, Decision Tree, Support Vector Machines(SVM)
    
Train SVM classifier using sklearn digits dataset (i.e. from sklearn.datasets import load_digits) and then,

Measure accuracy of your model using different kernels such as rbf and linear.

Tune your model further using regularization and gamma parameters and try to come up with highest accurancy score

Use 80% of samples as training data size

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.datasets import load_digits
digits=load_digits()

In [2]:
dir(digits)

# get data to a dataframe
df=pd.DataFrame=(digits.data)
df


array([[ 0.,  0.,  5., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ..., 10.,  0.,  0.],
       [ 0.,  0.,  0., ..., 16.,  9.,  0.],
       ...,
       [ 0.,  0.,  1., ...,  6.,  0.,  0.],
       [ 0.,  0.,  2., ..., 12.,  0.,  0.],
       [ 0.,  0., 10., ..., 12.,  1.,  0.]])

In [3]:
X=digits.data
y=digits.target
y

array([0, 1, 2, ..., 8, 9, 8])

In [4]:
# Train and test data split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2)

X_test

array([[ 0.,  0.,  0., ..., 13.,  1.,  0.],
       [ 0.,  0., 15., ...,  6.,  2.,  0.],
       [ 0.,  0.,  5., ..., 10.,  0.,  0.],
       ...,
       [ 0.,  0., 10., ..., 16., 15.,  0.],
       [ 0.,  0., 10., ...,  1.,  0.,  0.],
       [ 0.,  3., 15., ..., 13.,  3.,  0.]])

# Using Support Vector Machines(SVM): A Supervised Classification Algorithm

In [30]:
# SVM 

# TUNING HYPER-PARAMETERS

from sklearn.svm import SVC

# with default tuning parameters
model_svc1=SVC()

# run1 
model_svc1.fit(X_train,y_train)

#score
acc_svm1=model_svc1.score(X_test,y_test)
print(f"The accuracy of SVM1 model is:{round(acc_svm1*100,2)}%")

The accuracy of SVM1 model is:97.78%


In [51]:
model_svc2=SVC(C=100,kernel='linear')

# run2
model_svc2.fit(X_train,y_train)

#score
acc_svm2=model_svc2.score(X_test,y_test)
print(f"The accuracy of SVM2 model is:{round(acc_svm2*100,2)}%")

The accuracy of SVM2 model is:96.94%


In [58]:
model_svm3=SVC(C=100,kernel='poly')

# run2
model_svm3.fit(X_train,y_train)

#score
acc_svm3=model_svc3.score(X_test,y_test)
print(f"The accuracy of SVM3 model is:{round(acc_svm3*100,2)}%")
print('SVM3 is the best SVM model')

# get prediction from best svm model
y_pred_svm=model_svm3.predict(X_test)

The accuracy of SVM3 model is:98.33%
SVM3 is the best SVM model


SVM model with C=100 and kernel='rbf' or 'poly' gives maximum accuracy of 98.33%

In [79]:
# Evaluation metrics for best SVM Model
from sklearn.metrics import confusion_matrix, classification_report

cm_svm=confusion_matrix(y_test,y_pred_svm)
print(cm_svm)

# 
print(classification_report(y_test,y_pred_svm))

[[33  0  0  0  0  0  0  0  0  0]
 [ 0 40  0  0  0  0  0  0  0  0]
 [ 0  0 41  0  0  0  0  0  0  0]
 [ 0  0  0 30  0  0  0  0  0  0]
 [ 0  0  0  0 39  0  0  0  0  0]
 [ 0  0  0  0  0 42  0  0  0  1]
 [ 0  1  0  0  0  0 44  0  0  0]
 [ 0  0  0  0  0  0  0 22  0  0]
 [ 0  1  0  0  0  0  0  0 32  0]
 [ 0  0  0  0  0  1  0  1  1 31]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        33
           1       0.95      1.00      0.98        40
           2       1.00      1.00      1.00        41
           3       1.00      1.00      1.00        30
           4       1.00      1.00      1.00        39
           5       0.98      0.98      0.98        43
           6       1.00      0.98      0.99        45
           7       0.96      1.00      0.98        22
           8       0.97      0.97      0.97        33
           9       0.97      0.91      0.94        34

    accuracy                           0.98       360
   macro avg       

# Using Logistic Regression: Multi Class Logistic Regression

In [11]:
from sklearn.linear_model import LogisticRegression
model_lr=LogisticRegression()

# Train
model_lr.fit(X_train,y_train)

# predict
y_pred_lr=model_lr.predict(X_test)

# score
acc_lr=model_lr.score(X_test,y_test)
print(f"The accuracy of Logisitic Rgeression model is:{round(acc_lr*100,2)}%")

The accuracy of Logisitic Rgeression model is:97.5%


C:\Users\rashi\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [12]:
# Evaluation metrics for best SVM Model
from sklearn.metrics import confusion_matrix, classification_report

cm_lr=confusion_matrix(y_test,y_pred_lr)
print(cm_lr)

# 
print(classification_report(y_test,y_pred_lr))

[[34  0  0  0  0  0  0  0  0  0]
 [ 0 34  0  1  0  0  0  0  0  0]
 [ 0  0 33  0  0  0  0  0  0  0]
 [ 0  0  1 39  0  0  0  0  0  0]
 [ 0  0  0  0 35  0  0  0  0  0]
 [ 0  0  0  1  0 33  0  0  0  0]
 [ 0  0  0  0  0  1 41  0  0  0]
 [ 0  0  0  0  0  0  0 40  0  0]
 [ 0  2  0  0  0  0  0  0 31  1]
 [ 0  0  0  1  1  0  0  0  0 31]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        34
           1       0.94      0.97      0.96        35
           2       0.97      1.00      0.99        33
           3       0.93      0.97      0.95        40
           4       0.97      1.00      0.99        35
           5       0.97      0.97      0.97        34
           6       1.00      0.98      0.99        42
           7       1.00      1.00      1.00        40
           8       1.00      0.91      0.95        34
           9       0.97      0.94      0.95        33

    accuracy                           0.97       360
   macro avg       

# Decision Tree Algorithm

In [14]:
# Using Decision Tree

from sklearn import tree
model_dt=tree.DecisionTreeClassifier()

model_dt.fit(X_train,y_train)

DecisionTreeClassifier()

In [15]:
# score of model
acc_dt=model_dt.score(X_test,y_test)
print(f"The score of the Decision Tree model is: {round(acc_dt*100,2)}%")

y_pred_dt=model_dt.predict(X_test)

The score of the Decision Tree model is: 85.83%


In [16]:
# Confusion matrix
from sklearn.metrics import confusion_matrix, classification_report


cm_dt=confusion_matrix(y_test,y_pred_dt)
print(cm_dt)

# 
print(classification_report(y_test,y_pred_dt))

[[33  0  0  0  0  1  0  0  0  0]
 [ 0 30  0  1  1  0  0  0  3  0]
 [ 1  0 29  2  0  0  1  0  0  0]
 [ 0  1  0 31  0  0  0  1  4  3]
 [ 0  2  0  0 26  1  1  5  0  0]
 [ 0  1  0  0  2 31  0  0  0  0]
 [ 0  0  0  1  2  1 38  0  0  0]
 [ 0  2  0  0  2  0  0 35  0  1]
 [ 0  3  1  1  0  0  0  0 27  2]
 [ 0  0  0  1  0  2  0  0  1 29]]
              precision    recall  f1-score   support

           0       0.97      0.97      0.97        34
           1       0.77      0.86      0.81        35
           2       0.97      0.88      0.92        33
           3       0.84      0.78      0.81        40
           4       0.79      0.74      0.76        35
           5       0.86      0.91      0.89        34
           6       0.95      0.90      0.93        42
           7       0.85      0.88      0.86        40
           8       0.77      0.79      0.78        34
           9       0.83      0.88      0.85        33

    accuracy                           0.86       360
   macro avg       

# Random Forest Classifier


In [22]:
from sklearn.ensemble import RandomForestClassifier
model_rf=RandomForestClassifier(n_estimators=250)

# train model
model_rf.fit(X_train,y_train)

# prediction
y_pred_rf=model_rf.predict(X_test)

# score
acc_rf=model_rf.score(X_test,y_test)
print(f"The score of the Random Forest Classifier is: {round(acc_rf*100,2)}%")

The score of the Random Forest Classifier is: 98.89%


In [23]:
# Confusion matrix
from sklearn.metrics import confusion_matrix, classification_report


cm_rf=confusion_matrix(y_test,y_pred_rf)
print(cm_rf)

# 
print(classification_report(y_test,y_pred_rf))

[[39  0  0  0  0  0  0  0  0  0]
 [ 0 38  0  0  0  0  0  0  0  0]
 [ 0  0 37  0  0  0  0  0  0  0]
 [ 0  0  0 33  0  1  0  0  0  0]
 [ 0  0  0  0 38  0  0  1  0  0]
 [ 0  0  0  0  0 30  0  0  0  1]
 [ 0  0  0  0  0  0 37  0  0  0]
 [ 0  0  0  0  0  0  0 34  0  0]
 [ 0  0  0  0  0  0  0  0 34  0]
 [ 0  0  0  0  0  0  0  0  1 36]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        39
           1       1.00      1.00      1.00        38
           2       1.00      1.00      1.00        37
           3       1.00      0.97      0.99        34
           4       1.00      0.97      0.99        39
           5       0.97      0.97      0.97        31
           6       1.00      1.00      1.00        37
           7       0.97      1.00      0.99        34
           8       0.97      1.00      0.99        34
           9       0.97      0.97      0.97        37

    accuracy                           0.99       360
   macro avg       